codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/common/shapevisitor/conversionwriter/BaseConversionWriter.java

```diff
@@ -0,0 +1,107 @@
+ * shapes that can contain themselves as members (Unions and Structures). These shapes require
+ * delegating conversions to functions that recurse at runtime, and not at code generation time.
+ */
+public abstract class BaseConversionWriter {
+  // Store the set of shapes for which the subclass has already generated conversion methods
+  final Set<Shape> generatedShapes = new HashSet<>();
+  // Queue of shapes to generate
+  final List<Shape> shapesToGenerate = new ArrayList<>();
+  // Flag to block generating inside a file while already generating another function
+  // (prevents generating a conversion function inside another conversion function)
+  boolean generating = false;
+
+  protected GenerationContext context;
+  // Writer pointing to the in-code location where the ShapeVisitor calling the subclass is writing
+  protected PythonWriter writer;
+
+  /**
+   * Writes a function that converts from the AWS SDK dict to the Dafny-modelled shape. If the shape
+   * has members that also require writing a conversion function, this function will also write
+   * conversion methods for those shapes (recursively). If this method is called for the same shape
+   * multiple times, subsequent calls will not write a conversion method for the shape. This is the
+   * ONLY interface by which clients of this class' subclasses can request writing a new shape.
+   *
+   * @param shape
+   * @param context
+   * @param writer
+   */
+  public void baseWriteConverterForShapeAndMembers(
+      Shape shape, GenerationContext context, PythonWriter writer) {
+
+    this.context = context;
+    // Store where this is being written from where the original ShapeVisitor was dispatched (e.g.
+    // serialize, shim);
+    // This allows us to write imports in the original file
+    this.writer = writer;
+    // Do NOT write any converters for wrapped localServices.
+    // The wrapped localService should ONLY generate a Shim class.
+    // The Shim will use the converters generated as part of the localService.
+    if (context
+        .applicationProtocol()
+        .equals(
+            DafnyPythonWrappedLocalServiceProtocolGenerator
+                .DAFNY_PYTHON_WRAPPED_LOCAL_SERVICE_PROTOCOL)) {
+      return;
+    }
+
+    // Enqueue this shape if this class has not generated a converter for it or is not already in
+    // queue
+    if (!generatedShapes.contains(shape) && !shapesToGenerate.contains(shape)) {
+      shapesToGenerate.add(shape);
+    }
+
+    // If `writeConverterForShapeAndMembers` is called while already writing another converter,
+    // do NOT write the new converter inside the definition of the in-progress converter.
+    // `shape` will be picked up once the in-progress converter is finished.
+    while (!generating && !shapesToGenerate.isEmpty()) {
+      // Indicate to recursive calls that this class is writing a converter to
+      //   prevent writing multiple conversion methods at once
+      generating = true;
+
```

Comment:
There's a non zero chance for a race condition here (though unlikely for a manually executed codegen), we should use a MutexLock here.

Reply:
I may leave this out of scope if that's alright... I'm trying to avoid making significant changes at this point...

Reply:
Since this is not a direct prod code, I'm okay with this but we definitely NEED to add a comment saying this can happen and cut a SIM as well. No one wants to debug an issue related to a threaded race condition.

Reply:
Good call. https://sim.amazon.com/issues/CrypTool-5331